RFC Draft.md

```diff
@@ -0,0 +1,583 @@
+
+This RFC is broken down into sections covering:
+
+- Library-wide TypeScript improvements
+- Changes to some of our core utilities such as `Hub` and Amplify configuration
+- Specific improvements to the `Auth`, `Storage`, and `API` categories and associated APIs
+
+We're also requesting feedback on any other TypeScript issues or pain points that you may have encountered not explicitly covered in this RFC.
+
+# Library-wide TypeScript Improvements
+
+Amplify JS will be making the following improvements to our TypeScript support. These improvements will be applied across the entire library, not just the categories highlighted below.
+
+- `strict` typings — We will be applying `strict` mode to the entire library to improve the usability of our types.
+- Better runtime error typing — We will provide utilities for asserting type information of runtime errors emitted by Amplify.
+- Upgraded TypeScript version — We will be upgrading the version of TypeScript that Amplify uses and provide explicit type definitions for customers using older versions. This will provide a variety benefits such as removing the need to specify `skipLibCheck` when using Amplify with newer versions of TypeScript.
+
+# Utility Changes
+
+Amplify is proposing the following changes to our core utilities.
+
+## TypeScript support for Amplify Hub channels
+
+We are improving developer experience by adding strict type support to Hub channels, events, and payloads. An example of the developer experience when listening for `auth` events is highlighted below.
+
+**Amplify v5 & v6 (`aws-amplify@^5`)**
+
+```Typescript
+Hub.dispatch("auth", {
+  event: "signInFailure",
+  data: new AuthError("Sign-in failed"),
+});
+
+// Events and payload data are not inferred.
+Hub.listen(channel, ({ payload }) => {
+  switch (payload.event) {
+    case "signInFailure":
+      const data = payload.data;
+      break;
+  }
+});
+```
+
+**Amplify v5 DX**
+
+https://user-images.githubusercontent.com/70438514/222475294-9cf9f535-b717-4725-9e5a-7c362e74a054.mov
+
+<br />
+
+**Proposed Amplify v6 DX**
+
+https://user-images.githubusercontent.com/70438514/222512918-ea72b2c4-9568-45e6-86c6-3c5226820e7d.mov
+
+## TypeScript support for custom Hub channels
+
+We are improving developer experience by adding strict type support to custom Hub channels, events, and payloads.
+
+**Amplify v5 (`aws-amplify@5`)**
+
+```Typescript
+const customChannel = "custom_channel";
+const customEvent = "custom_event";
+const customData = "custom_data";
+
+Hub.dispatch(channel, {
+  event: customEvent,
+  data: customData
+});
+
+Hub.listen(channel, ({ payload }) => {
+  switch (payload.event) {
+    case customEvent:
+      const data = payload.data;
+      break;
+  }
+});
+```
+
+**Amplify v5 DX**
+
+https://user-images.githubusercontent.com/70438514/222475304-128de96a-9a06-4fcc-8f0d-8f19592ae9eb.mov
+
+<br />
+
+**Proposed Amplify v6 (`aws-amplify@6`)**
+
+```Typescript
+type customChannel = "custom_channel";
```

Comment:
Should these be PascalCased? I would guess these are customer defined, but it could be confusing to break convention